### PR TITLE
Added support for lambda expressions in `Callable` & Added `Defer` method for Godot.Object

### PR DESCRIPTION
**Pull request by lfshr**
Relates to https://github.com/godotengine/godot-proposals/issues/3474

This is more of a quality of life improvement for better readability. Adds the ability to create a callable from a lambda, and a `Defer` method on `Godot.Object` that wraps `new Callable(..).CallDeferred`.

Old
```csharp
public override void _Ready()
{
    _collisionShape = GetNode<CollisionShape2D>("Capsule");

    new Callable(new Action(() => _collisionShape.Disabled = true)).CallDeferred();
}
```

New
```csharp
public override void _Ready()
{
    _collisionShape = GetNode<CollisionShape2D>("Capsule");

    new Callable(() => _collisionShape.Disabled = true).CallDeferred();
    
    // OR 
    
    Defer(() => _collisionShape.Disabled = false);
}
```

### CODE REVIEW

**Comment by raulsntos** [comment]
This only works for lambdas that have no parameters and don't return a value. So it won't work in this case:

```csharp
public override void _Ready()
{
    _collisionShape = GetNode<CollisionShape2D>("Capsule");

    new Callable(disabled => _collisionShape.Disabled = disabled).CallDeferred(true);
    
    // OR 
    
	// Defer doesn't even have a way to provide parameters
    Defer(disabled => _collisionShape.Disabled = disabled);
}
```

Returns are ignored for deferred calls but they might be useful for non-deferred calls:

```csharp
bool isDisabled = (bool)new Callable(() => _collisionShape.Disabled).Call();
```

Although you wouldn't use a callable in this situation, it's just a very simple example.

**Comment by lfshr** [comment]
> This only works for lambdas that have no parameters and don't return a value. So it won't work in this case:
> 
> ```cs
> public override void _Ready()
> {
>     _collisionShape = GetNode<CollisionShape2D>("Capsule");
> 
>     new Callable(disabled => _collisionShape.Disabled = disabled).CallDeferred(true);
>     
>     // OR 
>     
> 	// Defer doesn't even have a way to provide parameters
>     Defer(disabled => _collisionShape.Disabled = disabled);
> }
> ```
> 
> Returns are ignored for deferred calls but they might be useful for non-deferred calls:
> 
> ```cs
> bool isDisabled = (bool)new Callable(() => _collisionShape.Disabled).Call();
> ```
> 
> Although you wouldn't use a callable in this situation, it's just a very simple example.

Lambdas in C# support capturing outside variables. 

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions#capture-of-outer-variables-and-variable-scope-in-lambda-expressions

```csharp
var disabled = true;
Defer(() => _collisionShape.IsDisabled = disabled)
```

A returning type is interesting. You can infer the returning type via something like:
```csharp
void Defer<TResult>(Expression<Func<TResult>> expression)
{
    ...
}
```

**Comment by lfshr** [comment]
> > This only works for lambdas that have no parameters and don't return a value. So it won't work in this case:
> > ```cs
> > public override void _Ready()
> > {
> >     _collisionShape = GetNode<CollisionShape2D>("Capsule");
> > 
> >     new Callable(disabled => _collisionShape.Disabled = disabled).CallDeferred(true);
> >     
> >     // OR 
> >     
> > 	// Defer doesn't even have a way to provide parameters
> >     Defer(disabled => _collisionShape.Disabled = disabled);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Returns are ignored for deferred calls but they might be useful for non-deferred calls:
> > ```cs
> > bool isDisabled = (bool)new Callable(() => _collisionShape.Disabled).Call();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Although you wouldn't use a callable in this situation, it's just a very simple example.
> 
> Lambdas in C# support capturing outside variables.
> 
> https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions#capture-of-outer-variables-and-variable-scope-in-lambda-expressions
> 
> ```cs
> var disabled = true;
> Defer(() => _collisionShape.IsDisabled = disabled)
> ```
> 
> A returning type is interesting. You can infer the returning type via something like:
> 
> ```cs
> TResult Defer<TResult>(Expression<Func<TResult>> expression)
> {
>     ...
> }
> ```
> 
> and consume it like:
> 
> ```cs
> var isDisabled = Defer(() => _collisionShape.IsDisabled);
> ```

That's a rubbish solution, you'd have to compile the expression. Could just use:
```csharp
TResult Defer<TResult>(Func<TResult> callback)
{
    return callback();
}
```

Doh!

**Comment by raulsntos** [comment]
> Lambdas in C# support capturing outside variables.

Yes, maybe that's good enough and we don't need to support every possible `Action` delegate.

> A returning type is interesting. You can infer the returning type via something like...

That won't work because the call is executed asynchronously (it's deferred) so you can't get the return value, `CallDeferred` always returns void. My issue was more with the non-deferred `Call` method of `Callable` since you would want to create that `Callable` from a `Func` instead of an `Action` in that case.

**Comment by lfshr** [comment]
So looking into this, I'm not sure adding this support is the best idea (or at least some investigation & discussion needs to happen). We're into the realms of compiler-generated `DisplayClass` classes.

In order to capture the context of a lambda a class is created by the compiler. All instances that need to be held onto until go in here (including value types) until the lambda is ready to release them; meaning stack allocation isn't possible. So the value type is boxed into a `DisplayClass`. This (especially in a game loop) will very quickly put pressure on Gen0 and be potentially detrimental to performance.

You can see some generated code here:
https://sharplab.io/#v2:D4AQTAjAsAULDOAXATgVwMaIAQGECGANgXgEYECmsA3rFnXQA7ICWAbnouVgCLkUDmHLgBM+Ablr1JdEAGZchYmXIAKEBDBZ0iknnQBrAJTSsNGPQt1RBLAF4tOvfrFTz9AL6xPcGHKzgsABVyJBxieHhqEz8SAHtYmwBJeG5meFIKYTssADNCeHIJGG9YAPUAdii3GQgANn8AFiwAWTxmADs1CAAGAG0AXSw8ZH54Y2rTEwt2ZCxhNIzyLPsUVEKp+hmsAFsATwD7dvIAdyCQxDC8CJVDIst6I9P8IkWVG7sAPh39sAA6ZNS6WUyzmC2BtxM3m8QA==

> Also, keep in mind that it would be very easy to just create that Defer method as an extension of Object without Godot having to provide it for you, so I'm also not sure it's super useful.

This is how the method started. It's perhaps a better idea to document that this is technically possible with a <here-be-dragons> warning.

**Comment by neikeq** [comment]
Honestly, `CallDeferred` makes the same or more allocations than the DisplayClass. It allocates the `params` array and boxes each argument. The DisplayClass doesn't box captured values.
I would like to tackle those issues in the future (hopefully in time for 4.0), but for now this addition is fine.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs [hunk 1]
```diff
@@ -132,6 +132,22 @@ public SignalAwaiter ToSignal(Object source, StringName signal)
...
+        /// <code>
+        /// public override void _Process(float delta)
+        /// {
+        ///     var collisionShape = GetNode&lt;CollisionShape2D&gt;("CollisionShape2D");
+        ///     Defer(() => collisionShape.Disabled = true);
+        /// }
+        /// </code>
+        /// </summary>
+        /// <param name="callback">Callback to call during idle time.</param>
+        protected void Defer(Action callback)
```
Comment (neikeq) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs:146]: I prefer to name this `CallDeferred` (to make it an overload), unless GDScript has a similar function/keyword called `defer`.

**Comment by neikeq** [comment]
By the way, we plan to switch to .NET 6 for Godot 4.0. Starting with C# 10, these constructors won't be needed any more as the language will be able to infer a natural delegate type for lambdas and method groups: https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/lambda-improvements.md

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs (hunk 1 above)
Comment (lfshr) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs:146]: The reason I changed it from `CallDeferred` was to make the distinction between the `public CallDeferred` and the protected lambda function. Happy to change it to an overload, should probably also make it public at that point. Opens up the opportunity of people doing things like `_myNode.CallDeferred(() => _myOtherNode.Disabled = true);`. Which gives me the fear 😆

**Comment by lfshr** [comment]
> By the way, we plan to switch to .NET 6 for Godot 4.0. Starting with C# 10, these constructors won't be needed any more as the language will be able to infer a natural delegate type for lambdas and method groups: https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/lambda-improvements.md

Oh! Good insight! So perhaps drop the `Callable(Action)` ctor. I still like the readability that comes with `Defer(() => ...)` (or `CallDeferred(() => ...)`

**Comment by lfshr** [comment]
> Honestly, `CallDeferred` makes the same or more allocations than the DisplayClass. It allocates the `params` array and boxes each argument. The DisplayClass doesn't box captured values. I would like to tackle those issues in the future (hopefully in time for 4.0), but for now this addition is fine.

This is something I don't have extensive knowledge about. If you're happy, I'm happy! 😄

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs (hunk 1 above)
Comment (neikeq) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs:146]: Shouldn't the method be static?

Comment (lfshr) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs:146]: @neikeq You're right; it probably should. Good catch.

So just to summarize changes I need to make:
1. Change `Defer` to `CallDeferred`?
2. Make method static.
3. Do we keep it `protected`, or is that too confusing when there are `public` overloads?
4. Do we want/need to support inferred returns as somebody mentioned? eg. `var isDisabled = CallDeferred(() => _colisionShape.IsDisabled)`? (I'm not convinced we do)

**Comment by raulsntos** [comment]
Since #64089 was merged the added constructor to `Callable` is no longer necessary. The existing `Callable(Delegate)` constructor already allows to do this in .NET 6.0:

```csharp
new Callable(() => _collisionShape.Disabled = true).CallDeferred();
```

We may want to consider adding a `Call` and `CallDeferred` overloads to `Object.base.cs` that take a `Delegate` as well.

This PR adds a `Defer` method that is essentially the `CallDeferred` overload I mention, so this PR could be salvaged to add these overloads instead. They should be `static`, take a `Delegate` parameter and return a `Variant` to be consistent with the generated methods (removing the `Variant` return from `CallDeferred` can be proposed separately and should be done in core first).

**Comment by neikeq** [comment]
I'm doing some changes to callables in order to make it reflection-free. I'd recommend waiting before making any further changes, as they could become obsolete.

**Comment by YuriSizov** [comment]
Seems like this is largely outdated by other changes, but there may be something to try and salvage. Closing then with the corresponding label.

**Pull request closed without merging**